src/passes/Outlining.cpp

```diff
@@ -0,0 +1,266 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "ir/utils.h"
+#include "pass.h"
+#include "stringify-walker.h"
+#include "wasm-builder.h"
+#include "wasm-stack.h"
+#include "wasm.h"
+
+namespace wasm {
+
+/*
+ * This walker performs a shallow visit of control-flow (try, if, block, loop)
+ * expressions and their simple expression siblings before then visiting the
+ * children of each control-flow expression in postorder. As a result, this
+ * walker un-nests nested control flow structures, so the expression visit order
+ * does not correspond to a normal postorder traversal of the function.
+ *
+ * For example, the below (contrived) wat:
+ * 1: (block
+ * 2:   (i32.add
+ * 3:     (i32.const 20)
+ * 4:     (i32.const 10))
+ * 5:   (if
+ * 6:     (i32.const 0)
+ * 7:     (then (return (i32.const 1)))
+ * 8:     (else (return (i32.const 0)))
+ * 9:   )
+ *
+ * Would have it's expressions visited in the following order (based on line
+ * number):
+ * 1, 3, 4, 2, 6, 5, 7, 8
+ *
+ * Of note:
+ *   - The add (line 2) binary operator's left and right children are visited
+ * first as they need to be on the stack before the add operation is executed
+ *   - The if-condition (i32.const 0) on line 6 is visited before the if
+ * expression
+ */
+template<typename SubType>
+void StringifyWalker<SubType>::walkModule(SubType* self, Module* module) {
+  self->wasm = module;
+  ModuleUtils::iterDefinedFunctions(*module, [&](Function* func) {
+    /*
+     * The ordering of the below lines of code are important. On each function
+     * iteration, we:
+     * 1. push a task for calling the handler function, to ensure that each
+     *    function has an opportunity to dequeue from StringifyWalker's
+     *    internally managed queue. This queue exists to provide a way for
+     *    control flow to defer scanning their children.
+     * 2. push a task for adding a unique symbol, so that after the function
+     *    body is visited as a single expression, there is a a separator between
+     *    the symbol for the entire function and the symbol for the function's
+     *    children
+     * 3. then we call walk, which will visit the function body as a single unit
+     * 4. finally we call addUniqueSymbol directly to ensure the string encoding
+     *    for each function is terminated with a unique symbol, acting as a
+     *    separator between each function in the program string
+     */
+    self->pushTask(StringifyWalker::handler, nullptr);
+    self->pushTask(StringifyWalker::addUniqueSymbol, &func->body);
+    self->walk(func->body);
+    self->addUniqueSymbol(self, &func->body);
```

Comment:
IIUC, this pushes two tasks, so the queue looks like this:
```
[handler, addUniqueSymbol]
```
Then we call walk, which first pushes the body,
```
[handler, addUniqueSymbol, body]
```
and then it pops the body and proceeds to do the walk. When it finishes with the body we are left with
```
[handler, addUniqueSymbol]
```
and then we pop `addUniqueSymbol`, and then pop `handler`.

Is that right?

If so, can't we call `addUniqueSymbol` and `handler` directly after the walk? I feel I'm missing something.

Reply:
completely refactored! take a look, I hope you like it :)